File: go/collatz.go
package main
// Collatz Conjecture
// Nate Jackson
// CSC330 F22

// Takes in two number from the command line, 
// calculates the length of each number's collatz sequence,
// then sorts on the lengths of the sequences and the integer values themselves
// displaying only the 10 largest lengths
import (
	"fmt"
	"os"
	"strconv"
)
//struct of two parallel arrays
//could not do array of structs like in other languages because of
// the way go does pointers, couldn't make it work
// but this works just as well
type Parallel struct {
	nums []int
	lengths []int
}

func main(){
	//command line input
	num1, err1 := strconv.Atoi(os.Args[1])
	if err1 != nil {
		panic(err1)
	}
	num2, err2 := strconv.Atoi(os.Args[2])
	if err2 != nil {
		panic(err2)
	}
	if num1 > num2 {
		temp := num1
		num1 = num2
		num2 = temp
	}
	var v  Parallel

	// filling arrays, first nums, then sequence lengths
	createArray(&v, num1, num2)
	for i := 0; i < len(v.nums); i++ {
		length := sequence(v.nums[i])
		v.lengths = append(v.lengths, length)
	}

	lengthSort(&v)
	// now that arrays are sorted, keep only top 10 values
	v.nums = v.nums[0:10]
	v.lengths = v.lengths[0:10]
	fmt.Println("Sorted based on sequence length:")
	for i := 0; i < len(v.nums); i++{
		fmt.Println(v.nums[i],"	", v.lengths[i])
	}

	integerSort(&v)
	fmt.Println("Sorted based on integer size:")
	for i := 0; i < len(v.nums); i++{
		fmt.Println(v.nums[i],"	", v.lengths[i])
	}

}

func createArray(v *Parallel,num1, num2 int){

	for i := num1; i <= num2; i++ {
		v.nums = append(v.nums, i)
	}
}
//calculates length of collatz sequence for given number
func sequence(num int) int{
	temp := num
	length := 0
	for temp != 1 {
		if temp % 2 == 0 {
			temp /= 2
		} else {
			temp = temp * 3 + 1
		}
		length += 1
	}
	return length
}
// bubble sort of both arrays in struct
func lengthSort(v *Parallel){
	// if duplicate length values, keep only smallest integer
	for i := 0; i < len(v.nums); i++ {
		for j := 0; j < len(v.nums); j++ {
			if v.nums[i] != v.nums[j] && v.lengths[i] == v.lengths[j] {
				v.lengths[j] = 0
			}
		}
	}

	for i := len(v.nums)-1; i >= 0; i-- {
		for j := 0; j < i; j++ {
			if v.lengths[j+1] >= v.lengths[j] {
				temp := v.lengths[j+1]
				v.lengths[j+1] = v.lengths[j]
				v.lengths[j] = temp
				temp = v.nums[j+1]
				v.nums[j+1] = v.nums[j]
				v.nums[j] = temp
			}
		} 
	}
}
//bubble sort based on integer values
//will only ever sort 10 valuess
func integerSort(v *Parallel){
	for i := len(v.nums)-1; i >= 0; i-- {
		for j := 0; j < i; j++ {
			if v.nums[j+1] >= v.nums[j] {
				temp := v.nums[j+1]
				v.nums[j+1] = v.nums[j]
				v.nums[j] = temp
				temp = v.lengths[j+1]
				v.lengths[j+1] = v.lengths[j]
				v.lengths[j] = temp
			}
		} 
	}
}

File: julia/collatz.jl
#!/usr2/local/julia-1.8.2/bin/julia
#  Collatz Conjecture
#  Nate Jackson
#  CSC330 F22

#  Takes in two number from the command line, 
#  calculates the length of each number's collatz sequence,
#  then sorts on the lengths of the sequences and the integer values themselves
#  displaying only the 10 largest lengths
mutable struct parallel
    num::Int32
    length::Int32
end
# calculates length of collatz sequence for given number
function sequence(num)
    length = 0
    while num != 1
        if num % 2 == 0
            num /= 2;
        else
            num = num * 3 + 1
        end
        length += 1
    end
    return length
end
# bubble sort based on sequence length
function lengthSort(v::Array{parallel})
	# if duplicate length values, keep only smallest integer
    for i in 1:length(v)
        for j in 1:length(v)
            if v[i].num != v[j].num && v[i].length == v[j].length
                v[j].length = 0
            end
        end
    end

    for i in length(v)-1:-1:1
        for j in 1:i
            if v[j+1].length >= v[j].length
                temp = v[j+1].length
                v[j+1].length = v[j].length
                v[j].length = temp
                temp = v[j+1].num
                v[j+1].num = v[j].num
                v[j].num = temp
            end
        end
    end
end
# bubble sort based on integer values
function integerSort(v::Array{parallel})
    for i in length(v)-1:-1:1
        for j in 1:i
            if v[j+1].num >= v[j].num
                temp = v[j+1].num
                v[j+1].num = v[j].num
                v[j].num = temp
                temp = v[j+1].length
                v[j+1].length = v[j].length
                v[j].length = temp
            end
        end
    end
end

# Main
num1 = parse(Int32, ARGS[1])
num2 = parse(Int32, ARGS[2])
if num1 > num2
    temp = num1
    num1 = num2
    num2 = temp
end
v = Array{parallel, 1}(undef, 51)

# filling arrays, first nums, then sequence lengths
for i in num1:num2
    index = i-num1+1
    s = parallel(i, sequence(i))
    v[index] = s
end

lengthSort(v)
# now that arrays are sorted, keep only top 10 values
v_final = v[1:1:10]
println("Sorted based on sequence length:")
for i in 1:length(v_final)
    print(v_final[i].num)
    print("\t")
    println(v_final[i].length)
end

integerSort(v_final)
println("Sorted based on integer value:")
for i in 1:length(v_final)
    print(v_final[i].num)
    print("\t")
    println(v_final[i].length)
end